pkg/vpcmodel/groupingSelfLoop.go

```diff
@@ -0,0 +1,335 @@
+	return mergeSelfLoops(toMergeCouples, groupingSrcOrDst, srcGrouping)
+}
+
+func (g *GroupConnLines) groupsToBeMerged(groupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool) (toMergeCouples [][2]string) {
+	toMergeCouples = make([][2]string, 0)
+	// the to be grouped src/dst in set representation, will be needed to compute potential groups to be merged
+	// and to compute the deltas
+	setsToGroup := createGroupingSets(groupingSrcOrDst, srcGrouping)
+	relevantKeys := relevantKeysToCompare(groupingSrcOrDst)
+	keyToMergeCandidates := g.mergeCandidates(groupingSrcOrDst, srcGrouping, setsToGroup, relevantKeys)
+
+	for _, key := range relevantKeys {
+		keyLines := groupingSrcOrDst[key]
+		//  is there a different line s.t. the keyLines were not merged only due to self loops?
+		// 	going over all couples of items: merging them if they differ only in self loop element
+		// mergeCandidates of a singleton 'key' are all lines in which the group contains 'key'
+		//  if key is not a singleton then mergeCandidates will be empty
+		mergeCandidates, ok := keyToMergeCandidates[key]
+		if !ok {
+			continue
+		}
+		for candidate := range mergeCandidates {
+			candidateLines := groupingSrcOrDst[candidate]
+			// delta between keyLines to candidateLines must be 0
+			mergeGroups := deltaBetweenGroupedConnLines(srcGrouping, keyLines, candidateLines, setsToGroup[key], setsToGroup[candidate])
+			// delta between the keyLines is 0 - merge keyLines
+			if mergeGroups {
+				toMergeCouples = append(toMergeCouples, [2]string{key, candidate})
+			}
+		}
+	}
+	return toMergeCouples
+}
+
+// a group is candidate to be merged only if it has only internal nodes
+func relevantKeysToCompare(groupingSrcOrDst map[string][]*GroupedConnLine) (relevantKeys []string) {
+	relevantKeys = make([]string, 0, len(groupingSrcOrDst))
+	for key, lines := range groupingSrcOrDst {
+		if lines[0].isSrcOrDstExternalNodes() {
+			continue
+		}
+		relevantKeys = append(relevantKeys, key)
+	}
+	return
+}
+
+// optimization to reduce the worst case of finding couples to merge from O(n^4) to O(n^2)
+// where n is the number of nodes.
+// a couple of []*GroupedConnLine is candidate to be merged only if:
+//  1. They are of the same connection
+//  2. If vsis, of the same subnet
+//  3. The src/dst is a singleton contained in the dst/src
+//     in one path on groupingSrcOrDst we prepare a map between each key to the keys that are candidate to be merged with it.
+//     Before the grouping there are at most O(n^20) lines of src -> dst
+//     The last condition implies that each original src -> dst (where src and dst are a single endpoint) can induce a single
+//     candidate (at most), and each singleton key have at most n candidates. Hence, there are at most
+//     O(n^3) merge candidate, which implies O(n^3) time complexity of groupsToBeMerged
+func (g *GroupConnLines) mergeCandidates(groupingSrcOrDst map[string][]*GroupedConnLine, srcGrouping bool,
```

Comment:
function name is confusing, this function computes candidates to be merged later, and does not really merge.

Reply:
renamed to `findMergeCandidates`